File: day1/py/main.py
def firstlast(s):
    for ch in s:
        if ch.isnumeric():
            first = ord(ch) - 0x30
            break
    for ch in reversed(s):
        if ch.isnumeric():
            last = ord(ch) - 0x30
            break
    return (first, last)            
#/def

def findrepl(s):
    digits = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
    repls = ["1", "2", "3", "4", "5", "6", "7", "8", "9"]
    for idx, digit in enumerate(digits):
        pos = s.find(digit)
        if pos >= 0:
            temp = list(s)
            temp[pos + 1] = repls[idx]
            s = "".join(temp)
    for idx, digit in enumerate(digits):
        pos = s.rfind(digit)
        if pos >= 0:
            temp = list(s)
            temp[pos + 1] = repls[idx]
            s = "".join(temp)
    return s
#/def

lines = open("../input.txt", "r").readlines()

#Part One
sum = 0
for line in lines:
    res = firstlast(line)
    sum += res[0] * 10 + res[1]
print("part one = ", sum)

#Part Two
sum = 0
for line in lines:
    newline = findrepl(line)
    res = firstlast(newline)
    sum += res[0] * 10 + res[1]
print("part two = ", sum)

File: day1/cpp/main.cpp
#include <ctype.h>
#include <stdio.h>
#include <string.h>

struct Digit {
    const char* str;
    int len;
};

struct Digit digits[] = {
    { "one", 3},
    { "two", 3},
    { "three", 5},
    { "four", 4},
    { "five", 4},
    { "six", 3},
    { "seven", 5},
    { "eight", 5},
    { "nine", 4}   
};

int main(int argc, char **argv)
{   
    if (argc < 2) {
        printf("usage: main.exe inputfile\n");
        return 1;
    }

    FILE *fp = fopen(argv[1], "rt");
    if (fp == NULL) {
        printf("can't open %s\n", argv[1]);
        return 1;
    }

//part One
    char line[256];
    int sum = 0;
    while (fgets(line, sizeof(line), fp)) {
        int firstNum = -1;
        int lastNum = -1;
        for (int i = 0; i < strlen(line); i++) {
            if (isdigit(line[i])) {
                if (firstNum < 0) {
                    firstNum = lastNum = line[i] - 0x30;
                } else {
                    lastNum = line[i] - 0x30;
                }
            }
        }
        sum += firstNum * 10 + lastNum;
    }
    printf ("part one sum = %d\n", sum);

//part Two
    sum = 0;
    fseek(fp, 0, SEEK_SET);
    while (fgets(line, sizeof(line), fp)) {
        int firstNum = -1;
        int lastNum = -1;
        for (int i = 0; i < strlen(line); i++) {
            if (isdigit(line[i])) {
                if (firstNum < 0) {
                    firstNum = lastNum = line[i] - 0x30;
                } else {
                    lastNum = line[i] - 0x30;
                }
            } else {
                for (int j = 0; j < (sizeof(digits) / sizeof(Digit)); j++) {
                    if (!strncmp(&line[i], digits[j].str, digits[j].len)) {
                        if (firstNum < 0) {
                            firstNum = lastNum = j + 1;
                        } else {
                            lastNum = j + 1;
                        }
                        break;
                    }
                }
            }
        }
        sum += firstNum * 10 + lastNum;
    }
    printf("part two sum = %d\n", sum);
    return 0;
}